### PR TITLE
offer to email one shopping list for all tickets, or just one ticket

### PR DESCRIPTION
**Pull request by leifwalsh**

### CODE REVIEW

File: functions/slackEvents.js [hunk 1]
```diff
@@ -139,10 +140,33 @@ async function onMessageEvent(event) {
...
+  if (!volunteer) {
+    throw new Error(`No volunteer found for slack id ${user.id}`);
+  }
+
   switch (action.action_id) {
   case 'email_shopping_list':
     try {
-      await emailShoppingList(action.value, user.id);
+      const records = await getRecordsWithTicketID(INTAKE_TABLE, action.value);
+      validateVolunteerAssigned(records, volunteerKey);
```
Comment (mlauter) [inline functions/slackEvents.js:152]: ah good call

File: functions/slackEvents.js [hunk 2]
```diff
@@ -139,10 +140,33 @@ async function onMessageEvent(event) {
...
+    }
+
+    return 0;
+  case 'email_consolidated_shopping_list':
+    try {
+      const records = await getRecordsWithFilter(INTAKE_TABLE, {
+        status: 'Assigned / In Progress',
+        deliveryVolunteer: volunteerKey
+      });
+      validateVolunteerAssigned(records, volunteerKey);
```
Comment (mlauter) [inline functions/slackEvents.js:167]: is this necessary here if that's how you're filtering the tickets from airtable

File: functions/slackEvents.js [hunk 3]
```diff
@@ -167,49 +191,86 @@ async function sendMessageToSlackResponseUrl(responseUrl, message) {
...
+      }),
+      ', '
+    );
+    throw new Error(`Volunteer ${volunteerKey} requested a shopping list for tickets they are not assigned to: ${mismatchedTicketIDs}`);
+  }
+}
+
+function renderTicketMetadata(fields) {
+  return `
+**Ticket ID:** ${fields.ticketID}<br/>
```
Comment (mlauter) [inline functions/slackEvents.js:211]: should we consider turning this stuff into a template?

File: functions/messages.js [hunk 4]
```diff
@@ -202,11 +202,20 @@ const renderDeliveryDM = (ticketID, deliveryDMContent, deliveryChannel) => (
...
             action_id: 'email_shopping_list',
             value: ticketID
+          },
+          {
+            type: 'button',
+            text: {
+              type: 'plain_text',
+              text: 'One list for all my tickets'
+            },
+            style: 'primary',
```
Comment (mlauter) [inline functions/messages.js:217]: Should we choose one of these buttons to be `primary` and use a different styling for the other?

File: functions/slackEvents.js (hunk 2 above)
Comment (leifwalsh) [reply functions/slackEvents.js:167]: Probably not, just paranoid.

File: functions/slackEvents.js (hunk 3 above)
Comment (leifwalsh) [reply functions/slackEvents.js:211]: I was trying not to spend too long on it last night, but yes we should.

File: functions/messages.js (hunk 4 above)
Comment (leifwalsh) [reply functions/messages.js:217]: SGTM

Comment (leifwalsh) [reply functions/messages.js:217]: 2020-08-07-212236_1920x1080_scrot
How does that look?

**Comment by leifwalsh** [comment]
@mlauter I also figured out a way to deal with the max section length: we can submit multiple sections, each one by itself just needs to be under 3000 characters.

File: functions/messages.js [hunk 5]
```diff
@@ -179,14 +179,19 @@ _${safetyReminder}_
 
 :heart: :heart: :heart:`;
 
-  return content;
+  // We split the content into two separate sections because each section has a
+  // max length of 3000 characters, and some tickets have long delivery notes or
+  // things that cause us to overflow that. Breaking the epilogue into a
+  // separate section makes it more likely that both sections will be under 3000
+  // characters.
+  return [content, epilogue];
```
Comment (mlauter) [inline functions/messages.js:187]: 👍

**Pull request merged**